internal/handlers/post_handler.go

```diff
@@ -121,6 +127,151 @@ func (handler *PostHandler) CreatePost(w http.ResponseWriter, r *http.Request) {
 	}, http.StatusCreated)
 }
 
+// DeletePost Deletes a post with the given id
+func (handler *PostHandler) DeletePost(w http.ResponseWriter, r *http.Request) {
+	// Begin a new transaction
+	tx, transactionCtx, cancel := utils.BeginTransaction(w, r, handler.DatabaseManager.GetPool())
+	if tx == nil || transactionCtx == nil {
+		utils.WriteAndLogError(w, schemas.DatabaseError, http.StatusInternalServerError, nil)
+		return
+	}
+	var err error
+	defer utils.RollbackTransaction(w, tx, transactionCtx, cancel, err)
+
+	// Get the post ID from the URL
+	postId := chi.URLParam(r, utils.PostIdParamKey)
+
+	// Get the user ID from the JWT token
+	claims := r.Context().Value(utils.ClaimsKey).(jwt.MapClaims)
+
+	// Check if the user is the author of the post
+	queryString := "SELECT author_id FROM alpha_schema.posts WHERE post_id = $1"
+	row := tx.QueryRow(transactionCtx, queryString, postId)
+
+	var authorId string
+	if err := row.Scan(&authorId); err != nil {
+		if errors.Is(err, pgx.ErrNoRows) {
+			utils.WriteAndLogError(w, schemas.PostNotFound, http.StatusNotFound, err)
+			return
+		}
+
+		utils.WriteAndLogError(w, schemas.DatabaseError, http.StatusInternalServerError, err)
+		return
+	}
+
+	userId := claims["sub"].(string)
+	if userId != authorId {
+		utils.WriteAndLogError(w, schemas.DeletePostForbidden, http.StatusForbidden, errors.New("user is not the author of the post"))
+		return
+	}
+
+	// Delete the hashtags associated with the post, if the hashtag isn't used also delete it
+	queryString = `DELETE FROM alpha_schema.many_posts_has_many_hashtags 
+       				WHERE post_id_posts = $1 RETURNING hashtag_id_hashtags`
+	rows, err := tx.Query(transactionCtx, queryString, postId)
+	if err != nil {
+		utils.WriteAndLogError(w, schemas.DatabaseError, http.StatusInternalServerError, err)
+		return
+	}
+
+	queries := make([]string, 0)
+	hashtagIds := make([]string, 0)
+
+	for rows.Next() {
+		var hashtagId string
+		if err := rows.Scan(&hashtagId); err != nil {
+			utils.WriteAndLogError(w, schemas.DatabaseError, http.StatusInternalServerError, err)
+			return
+		}
+
+		queryString = `DELETE FROM alpha_schema.hashtags WHERE hashtag_id = $1 
```

Comment:
Eigentlich sollte aufgrund der Cascading Regel in der post tabelle alle einträge aus der Tabelle "posts_has_many_hashtags" gelöscht werden, wenn ein Post gelöscht wird. Man könnte dann später ein extra check einbauen, ob der Hashtag nicht mehr referenziert wird um ihn schließlich zu löschen. Man könnte auch einfach automatisch versuchen den Hashtag in der nächsten query zu löschen und den Fehler abzufangen, dass der Hashtag noch referenziert wird. Dann benötigen wir nur 2 SQL Statements

Reply:
True, hatte ich gar nicht bedacht. Baue ich noch um